### PR TITLE
Adapt (S)MC-ABC API to latest `sbi` version

### PR DESCRIPTION
**Pull request by psteinb**
the mcabc API changed with `sbi` 0.17.1 ... it occurred to me that this was not propagated correctly to `sbibm`

### CODE REVIEW

**Comment by psteinb** [comment]
btw, I could add a unit test for the `run` method to this PR. I saw there is none at the moment.

**Comment by psteinb** [comment]
See als https://github.com/mackelab/sbi/blob/17591339b47854c57e999e6fe4714d073379a58f/tests/abc_test.py#L53

**Comment by janfb** [comment]
thanks for fixing this! LGTB @jan-matthis

**Comment by jan-matthis** [comment]
Great, thanks for fixing this!

Regarding ABC tests, I'm considering doing that in the `sbi` repo only, now that we moved them there. What do you think?

**Comment by psteinb** [comment]
Sure thing, I just had in mind to add an interface test, that is all. Just in case `sbibm` gets in trouble with `sbi` versions 
in the future.

**Comment by jan-matthis** [comment]
I see, yes -- would be good to have a small test to check whether it runs at all (without asserting correctness of results etc.)

**Comment by psteinb** [comment]
Cool, I’ll try to do that tomorrow.

**Comment by psteinb** [comment]
I added some simple code, which produces an error from
```
#...
tests/algorithms/sbi/test_mcabc.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sbibm/algorithms/sbi/mcabc.py:113: in run
    log_prob_true_parameters = kde_posterior.log_prob(true_parameters.squeeze())
sbibm-main-venv/lib64/python3.9/site-packages/sbi/utils/kde.py:31: in log_prob
    self.kde.score_samples(parameters_unconstrained.numpy()).astype(np.float32)
sbibm-main-venv/lib64/python3.9/site-packages/sklearn/neighbors/_kde.py:231: in score_samples
    X = self._validate_data(X, order="C", dtype=DTYPE, reset=False)
sbibm-main-venv/lib64/python3.9/site-packages/sklearn/base.py:561: in _validate_data
    X = check_array(X, **check_params)

#...
                # If input is 1D raise error
                if array.ndim == 1:
>                   raise ValueError(
                        "Expected 2D array, got 1D array instead:\narray={}.\n"
                        "Reshape your data either using array.reshape(-1, 1) if "
                        "your data has a single feature or array.reshape(1, -1) "
                        "if it contains a single sample.".format(array)
                    )
E                   ValueError: Expected 2D array, got 1D array instead:
E                   array=[ 0.73016036 -0.3217839 ].
E                   Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

```
see [`74d92a3` (#30)](https://github.com/sbi-benchmark/sbibm/pull/30/commits/74d92a36695852ac31e0952995060a9e84c8015e)

File: tests/algorithms/sbi/test_mcabc.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    ],
+)
+def test_run_posterior_interface(
+    task_name,
+    num_simulations=100,
+    num_samples=5,
+):
+    task = sbibm.get_task(task_name)
+
+    samples = run_posterior(
```
Comment (janfb) [inline tests/algorithms/sbi/test_mcabc.py:26]: the run script returns a tuple: `(samples, num_simulations, )`

**Comment by janfb** [comment]
regarding the error: this is a `transform` bug in `sbi`. will fix it and make a new release. thanks for catching this, yet another point for making these unit tests!

**Comment by psteinb** [comment]
Ok, then let's wait for `sbi` to fix that.

**Comment by janfb** [comment]
This is fixed now in `sbi`. Here, the `.squeeze()` of the `true_parameters` needs to be removed for this to work: the `log_prob` methods wants batched parameters, as common in `torch`. So this needs to be fixed both in `mcabc` and `smcabc`.

**Comment by janfb** [comment]
the pin to `sbi 0.17.2.` is merged into `main` already, so if you rebase on it and reinstall it should all be up to date.

**Comment by jan-matthis** [comment]
>  Here, the .squeeze() of the true_parameters needs to be removed for this to work: the log_prob methods wants batched parameters, as common in torch. So this needs to be fixed both in mcabc and smcabc.

To clarify, these are the respective lines:
- https://github.com/sbi-benchmark/sbibm/blob/main/sbibm/algorithms/sbi/mcabc.py#L113
- https://github.com/sbi-benchmark/sbibm/blob/main/sbibm/algorithms/sbi/smcabc.py#L162

**Comment by psteinb** [comment]
I discovered some minor things not working in `smcabc`. Should be ready for review with [`4d129e0` (#30)](https://github.com/sbi-benchmark/sbibm/pull/30/commits/4d129e082c77af8a50ddbeb06c6bc312eeac0380)

File: tests/algorithms/sbi/test_smcabc.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import pytest
```
Comment (janfb) [inline tests/algorithms/sbi/test_smcabc.py:1]: this replicates the testing code for `mcabc` almost 1-1. I would therefore vote for writing one test, e.g., `test_abc_run_scripts` or so, that makes the abc method an argument and parmaterizes it with pytest. E.g., 
```
from sbibm.algorithms.sbi.smcabc import run as run_smcabc
from sbibm.algorithms.sbi.mcabc import run as run_mcabc

...
@pytest.mark.parametrize("algorithm", (run_mcabc, run_smcabc))
...
def test_abc_run_scripts(task, algorithm, ...):
...
```

File: tests/algorithms/sbi/test_smcabc.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    samples, _, _ = run_posterior(
+        task=task,
+        num_simulations=num_simulations,
+        num_observation=3,
+        num_samples=num_samples,
+    )
+
+    # we are not interested in testing for correctness
+    assert len(samples.shape) > 0
```
Comment (janfb) [inline tests/algorithms/sbi/test_smcabc.py:34]: maybe test for number of samples here using `sample.shape == torch.Size(num_samples, task.dim_parameters)`?

**Review by janfb — CHANGES_REQUESTED** [review]
great, thanks for fixing all those bugs! I made some small comments, looks good otherwise.

File: tests/algorithms/sbi/test_smcabc.py (hunk 2 above)
Comment (psteinb) [reply tests/algorithms/sbi/test_smcabc.py]: fair enough, done with 08f8324

File: tests/algorithms/sbi/test_abc_run_methods.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+import sbibm
+from sbibm.algorithms.sbi.mcabc import run as run_mcabc
+from sbibm.algorithms.sbi.smcabc import run as run_smcabc
+from sbibm.metrics.c2st import c2st
+
+
+@pytest.mark.parametrize(
+    "task_name",
+    [tn for tn in ["gaussian_linear", "two_moons"]],
```
Comment (janfb) [inline tests/algorithms/sbi/test_abc_run_methods.py:12]: maybe shorten this with`("gaussian_linear", "two_moons")` instead of the list comprehension?

**Review by janfb — CHANGES_REQUESTED** [review]

File: tests/algorithms/sbi/test_smcabc.py (hunk 3 above)
Comment (psteinb) [reply tests/algorithms/sbi/test_smcabc.py]: And squashed another set of bugs! Done with cdd0439

File: tests/algorithms/sbi/test_abc_run_methods.py (hunk 4 above)
Comment (psteinb) [reply tests/algorithms/sbi/test_abc_run_methods.py]: :facepalm: thanks for spotting this. 1800192

File: sbibm/algorithms/sbi/mcabc.py [hunk 5]
```diff
@@ -105,12 +105,12 @@ def run(
 
     if kde:
         kde_posterior = output
-        samples = kde_posterior.sample(num_simulations)
+        samples = kde_posterior.sample(num_samples)
```
Comment (janfb) [inline sbibm/algorithms/sbi/mcabc.py:108]: good catch!

File: tests/algorithms/sbi/test_abc_run_methods.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import sbibm
+from sbibm.algorithms.sbi.mcabc import run as run_mcabc
+from sbibm.algorithms.sbi.smcabc import run as run_smcabc
+from sbibm.metrics.c2st import c2st
+
+
+@pytest.mark.parametrize(
+    "task_name",
+    ["gaussian_linear", "two_moons"],
```
Comment (janfb) [inline tests/algorithms/sbi/test_abc_run_methods.py:12]: we probably get it in one line if you remove the comma at the end.

**Review by janfb — APPROVED** [review]
Looks good, thanks! 
Good to go if all tests are passing :)

File: tests/algorithms/sbi/test_abc_run_methods.py (hunk 6 above)
Comment (psteinb) [reply tests/algorithms/sbi/test_abc_run_methods.py]: Just for you, @janfb 
[`f3d8ed5` (#30)](https://github.com/sbi-benchmark/sbibm/pull/30/commits/f3d8ed57292748d5b3b7fee7c22c2dc2624c2f0b)

**Comment by jan-matthis** [comment]
Tests are passing now 🚀

**Pull request merged**